### DISCUSSION TITLE
processing CosmoSkyMed Second Generation with StripmapApp

### DISCUSSION THREAD

**Discussion by sandrolr** [isce-framework/isce2#888] [Q&A]
Dear all,

I'm considering to task acquisitions from CosmoSkymed Second Generation.  Is it possible to process them using StripmapApp in isce2? Thanks  a lot in advance. 

Cheers,,

Alessandro

**Comment by piyushrpt** [comment]
In theory, minor changes (if any) to the reader should be sufficient make it work. However, there is not sufficient CSK 2nd gen data in the public domain for someone to have tested this extensively or even make the necessary changes to the reader.

**Comment by frx08** [comment]
Is it possible to get help in implementing the reader?
this is sample data from ESA server:
https://earth.esa.int/eogateway/ftp/missions/sample-data/third-party-missions/cosmo-skymed/second-generation/stripmap-himage/scs-b/Level-SCS-B-hdf5-HH.zip

when trying to use `unpackFrame_CSK.py`

I get the following error:
```
Traceback (most recent call last):
  File "/root/miniconda3/envs/isce_env/share/isce2/stripmapStack/unpackFrame_CSK.py", line 65, in <module>
    unpack(inps.h5dir, inps.slcdir)
  File "/root/miniconda3/envs/isce_env/share/isce2/stripmapStack/unpackFrame_CSK.py", line 42, in unpack
    obj.extractImage()
  File "/root/miniconda3/envs/isce_env/lib/python3.10/site-packages/isce/components/isceobj/Sensor/COSMO_SkyMed_SLC.py", line 239, in extractImage
    self.parse()
  File "/root/miniconda3/envs/isce_env/lib/python3.10/site-packages/isce/components/isceobj/Sensor/COSMO_SkyMed_SLC.py", line 111, in parse
    self.populateMetadata(fp)
  File "/root/miniconda3/envs/isce_env/lib/python3.10/site-packages/isce/components/isceobj/Sensor/COSMO_SkyMed_SLC.py", line 120, in populateMetadata
    self._populateInstrument(file)
  File "/root/miniconda3/envs/isce_env/lib/python3.10/site-packages/isce/components/isceobj/Sensor/COSMO_SkyMed_SLC.py", line 146, in _populateInstrument
    rangePixelSize = file['S01/SBI'].attrs['Column Spacing']
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "/root/miniconda3/envs/isce_env/lib/python3.10/site-packages/h5py/_hl/group.py", line 357, in __getitem__
    oid = h5o.open(self.id, self._e(name), lapl=self._lapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5o.pyx", line 190, in h5py.h5o.open
KeyError: "Unable to open object (object 'SBI' doesn't exist)"
```

**Reply by piyushrpt** [reply]
You should be able to use `h5py` to figure out what the right group names in the file and then update the reader to detect which flavor of CSK it is and use the right group names for parsing data / metadata.

**Reply by frx08** [reply]
I found that `unpackFrame_CSK_raw.py` accesses right group names ('S01/B001') using sensor `COSMO_SKYMED`
I updated the `extract_csk.c` lib to debug better (attached) and I get the following errors:

```
Opening file: CSG/CSG_SSAR1_SCS_B_0101_STR_007_HH_RA_F_20220905044559_20220905044605_1_F_42N_Z33_N00.h5
Extracting LUT...
HDF5-DIAG: Error detected in HDF5 (1.10.7) thread 1:
  #000: ../../../src/H5Adeprec.c line 202 in H5Aopen_name(): can't open attribute: 'Analog Signal Reconstruction Levels'
    major: Attribute
    minor: Can't open object
  #001: ../../../src/H5Aint.c line 556 in H5A__open_by_name(): unable to load attribute info from object header
    major: Attribute
    minor: Unable to initialize object
  #002: ../../../src/H5Oattribute.c line 511 in H5O__attr_open_by_name(): can't locate attribute: 'Analog Signal Reconstruction Levels'
    major: Attribute
    minor: Object not found
Failed to open LUT attribute.
Failed to extract LUT data.
```

so I made a dump for the CSG H5 (attached)
Is possible to get help in reading the attributes needed for the SLC conversion?

[CSG_H5.dump.txt](https://github.com/user-attachments/files/17932799/CSG_H5.dump.txt)
[extract_csk_v2.c.txt](https://github.com/user-attachments/files/17932800/extract_csk_v2.c.txt)

**Reply by piyushrpt** [reply]
If there is something different in the way data is organized, you are probably better off making a new sensor object. Since, there is not enough data in public domain again - I don't think others are going to step in here.  You can probably read up the product specification for the 2nd generation products and figure out what the new field names / changes are. There is really no need to use the C extension, you can do everything you need in h5py - it could be a few seconds slower but might be enough for your use case.

**Reply by svenborgstrom** [reply]
Hi all,
I wrote a post on the same subject (unpackFrame_CSK.py "GSG_data") on Dec. 11, 2022, with the same result as Francesco (KeyError: "Unable to open object (object 'SBI' doesn't exist)").
I'm interested in getting the CGS reader too, as I reguarly process CSK data (1st gen.) in my daily work.
I have no skills in setting up/managing scripts as I'm just an InSAR user, but I'm an Italian researcher with good contacts with ASI.
Piyush, If you think getting CSG data in public domain from ASI could help to solve the problem in setting up the reader, I can ask ASI if they can provide these data.
Just let me know.
Best,
Sven.
